### PR TITLE
`OutputSpec` for `DerivationGoal` and `DerivedPath`, today's `OutputSpec` -> `ExtendedOutputSpec`

### PR DESCRIPTION
**Pull request by Ericson2314**
`DerivedPath::Built` and `DerivationGoal` were previously using a regular `std::set<std::string<` with the convention that the empty set means all outputs. But it is easy to forget about this rule when processing those sets. Using `OutputsSpec` forces us to get it right.

I did this because of the deduplication @edolstra requested in https://github.com/NixOS/nix/pull/4543#discussion_r922164687. We can't do much good code reuse when `DerivedPath` doesn't use `OutputsSpec` like it should.

We need to split `OutputsSpec` from `ExtendedOutputsSpec` because the "default" case only makes sense for high-level installables with an `outputsToInstall` field. For `DerivationGoal`, and `DerivedPath` itself, only concrete sets and the "all" wild-card make sense.

### CODE REVIEW

**Comment by Ericson2314** [comment]
Hooray it is fixed!

**Comment by Ericson2314** [comment]
@edolstra I rebased and in the final 4th commit cleaned up the parser, following up from #4543. This is ready for review!

File: src/libcmd/installables.cc
Comment (edolstra) [inline src/libcmd/installables.cc:521]: Is this variable used anywhere (other than in the `if`)?

File: src/libcmd/installables.cc
Comment (edolstra) [inline src/libcmd/installables.cc:655]: Idem.

File: src/libstore/outputs-spec.cc
Comment (roberth) [inline src/libstore/outputs-spec.cc:9]:
```suggestion
bool OutputsSpec::wants(const std::string & outputName) const
```

`contains` suggests that something exists, which we may not know yet. `wants` does not suggest this.

File: src/libstore/build/derivation-goal.cc
Comment (edolstra) [inline src/libstore/build/derivation-goal.cc:1330]: This  is very similar to the code above (`realWantedOutputs`), so maybe this could be merged into a method on `OutputSpec`, e.g. `outputSpec.names_or(...)`.

File: src/libstore/build/derivation-goal.cc
Comment (tomberek) [inline src/libstore/build/derivation-goal.cc:1325]: presumably this is to minimize changes to the rest of this function. Should there be a TODO here?

File: src/libstore/misc.cc
Comment (edolstra) [inline src/libstore/misc.cc:304]: Maybe rename `evalStore0` to `evalStore_`. That's what we usually do for shadowed names.

Comment (edolstra) [reply src/libstore/misc.cc:304]: Actually why does `evalStore` need to be passed as a pointer instead of a reference?

File: src/libstore/outputs-spec.cc
Comment (edolstra) [inline src/libstore/outputs-spec.cc:16]: `return outputNames.count(outputName);`

File: src/libstore/outputs-spec.cc
Comment (edolstra) [inline src/libstore/outputs-spec.cc:98]: I wonder if this is strictly legal C++...

File: src/libstore/outputs-spec.cc
Comment (edolstra) [inline src/libstore/outputs-spec.cc:88]: Maybe this can just return a new `OutputsSpec`?

File: src/libcmd/installables.cc
Comment (edolstra) [inline src/libcmd/installables.cc:836]: What does `{}` as the outputs mean? I guess `AllOutputs`, but's it probably better to say that explicitly.

**Review by edolstra — APPROVED** [review]

File: src/libstore/outputs-spec.cc
Comment (roberth) [reply src/libstore/outputs-spec.cc:88]: fwiw, during the review, equality was mentioned, but this bool evaluates a subset relation instead.

File: src/libcmd/installables.cc
Comment (roberth) [reply src/libcmd/installables.cc:836]: Can we delete that constructor?

Comment (Ericson2314) [reply src/libcmd/installables.cc:836]: Ah! Good catch. So an issue is that `OutputNames` is a `typdef` and not it's own struct, so I am worried what is actually happening is that this is constructing an empty `std::set` which is then converted to an `OutputSpec`. In other words, I think this is a bug! I have at least seen this same bug developing the patch.

I think it would be better to make `OutputNames` its own struct (and we then yes can delete that constructor) to prevent this bug from sneaking in again. But I didn't do that because it makes some things, yes, more verbose.

File: src/libstore/misc.cc
Comment (Ericson2314) [reply src/libstore/misc.cc:304]: In the header it's has a default argument: `evalStore = nullptr`. I would like to do `evalStore = *this` and make it a reference but I don't think that is valid C++?

File: src/libstore/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/outputs-spec.cc:88]: For the record it is just used in one spot (I thought it might be used in more). I could change it or I could just inline it again as it was before whatever you all prefer.

File: src/libstore/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/outputs-spec.cc:9]: I have been thinking of `OutputSpec` as a merely the "finite power set of string ∪ ⊤", and so "contains" meant set contains for this notion of set of strings.

It is true that `OutputSpec` is usually *used* to represent the set of wanted outputs, but I feel a little odd calling the methods because I don't think it is the `OutputSpec` *itself* that does the wanting, it is merely used as means to convey something else's wants.

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting 2022-12-23:

- discussed how this fits with #6507
 - assigned @roberth for review

File: src/libstore/build/derivation-goal.cc
Comment (Ericson2314) [reply src/libstore/build/derivation-goal.cc:1325]: Hehe I thought the same, but I think this is actually the natural way for this one specific check --- which is "did we refer to any outputs that don't exist?"

File: src/libstore/tests/outputs-spec.cc
Comment (roberth) [inline src/libstore/tests/outputs-spec.cc:63]: Let's also test

```c++
ExtendedOutputsSpec::parse("*^bin")
ExtendedOutputsSpec::parse("*^*")
ExtendedOutputsSpec::parse("*^")
ExtendedOutputsSpec::parse("^*")
ExtendedOutputsSpec::parse("^")
ExtendedOutputsSpec::parse("bin^")
ExtendedOutputsSpec::parse("^bin")
ExtendedOutputsSpec::parse("drv!bin")
ExtendedOutputsSpec::parse("**")
```

"nonsensical" as they may be

File: src/libstore/outputs-spec.hh
Comment (roberth) [inline src/libstore/outputs-spec.hh:70]: Not a big fan of `print` for a method that doesn't perform I/O.
The following aligns with existing code

```suggestion
    std::string to_string() const;
```

File: src/libstore/outputs-spec.hh
Comment (roberth) [inline src/libstore/outputs-spec.hh:78]: These must be unit tested.

**Review by roberth — CHANGES_REQUESTED** [review]
Some comments and suggestions for testing. I'm not too clever and I have the memory of a goldfish, so we have to do better with testing. This code in particular would benefit from unit tests, for the obvious reason I just stated, but also as a way to "document" the json interface, which should be stable. Lastly testing may reveal aspects that we might not otherwise think of.

Unofficial checklist.

 - [x] is the idea good? has it been discussed by the Nix team?
 - [x] unit tests
 - [ ] functional tests (`tests/**.sh`)
   - reused
 - [ ] documentation in the manual
    - is this purely refactoring? yes. was the documentation for outputs up to par? might be best to separate that out; scoped out for now
 - [x] documentation in the code (if necessary; ideally code is already clear)
    - will judge when the tests are written
 - [x] documentation in the commit message (why was this change made? for future reference when maintaining the code)
 - [x] n/a documentation in the changelog (to announce features and fixes to existing users who might have to do something to finally solve their problem, and to summarize the development history)

File: src/libstore/tests/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/tests/outputs-spec.cc:23]: Test that they fail?

Comment (roberth) [reply src/libstore/tests/outputs-spec.cc:23]: I think so? Keeping the language small is a safe choice.

File: src/libstore/outputs-spec.hh
Comment (roberth) [reply src/libstore/outputs-spec.hh:78]: Doesn't have to be fancy; mostly to pin down the format.

**Comment by Ericson2314** [comment]
I taught myself in https://github.com/NixOS/nix/pull/7543 how to write the tests, I'll return to this once that one is merge (they overlap slightly, so conflicts either way).

File: src/libcmd/installables.cc
Comment (Ericson2314) [reply src/libcmd/installables.cc]: I rewrote this logic to hopefully be much easier to read.

File: src/libcmd/installables.cc
Comment (Ericson2314) [reply src/libcmd/installables.cc]: I rewrote this logic to hopefully be much easier to read. There was some stuff that just seemed redundant? I deleted it but do check.

File: src/libstore/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/outputs-spec.cc:98]: @tfc? I am not sure.

Comment (tfc) [reply src/libstore/outputs-spec.cc:98]: It's not illegal, but I would certainly consider it a dangerous pattern.
Is it maybe possible to refactor this so we have a merge function that is not a member of one of the to-be-merged items?

File: src/libstore/outputs-spec.hh
Comment (tfc) [inline src/libstore/outputs-spec.hh:11]: the STL container classes are *not* meant to be inherited from. Please don't.

File: src/libcmd/installables.cc
Comment (Ericson2314) [reply src/libcmd/installables.cc]: They are removed.

File: src/libstore/tests/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/tests/outputs-spec.cc]: Actually most of these do parse, because this really doesn't care what is to the right of a `^`.

File: src/libstore/outputs-spec.hh
Comment (Ericson2314) [reply src/libstore/outputs-spec.hh:11]: Is it a problem when it's a "newtype" like class --- exactly one superclass/field and so hopefully the same representation?

Comment (tfc) [reply src/libstore/outputs-spec.hh:11]: Hm... I see that you try to make it harder to create an empty set. 
Unfortunately I wouldn't deem this approach to be very effective since you can still write `auto myEmptyInstance = OutputNames({});` and it does not even look wrong, because the typename `OutputNames` does not convey that it's meant to never be empty. For this reason, one still has to put the same effort into the review to find out if we really never construct non-empty sets.

From a design perspective, this is not a case for inheritance.
I suggest to either drop this subclass which is from a design perspective more harm than good, or implement a class `template <typename T> class NonEmptySet` that *has* an `std::set` and that really implements the semantics of a non-empty data structure.

File: src/libstore/outputs-spec.hh
Comment (tfc) [inline src/libstore/outputs-spec.hh:28]: I also suggest to not subclass this standard class. Looking over the code, it does not seem to bring much advantage.

File: src/libstore/outputs-spec.hh
Comment (tfc) [inline src/libstore/outputs-spec.hh:32]: And here we have another subclass of STL standard classes... please, please, don't.

File: src/libstore/outputs-spec.hh
Comment (tfc) [inline src/libstore/outputs-spec.hh:65]: same pattern here. I have seen many code bases where developers subclassed stl classes - none of them aged well.

File: src/libstore/outputs-spec.hh
Comment (Ericson2314) [reply src/libstore/outputs-spec.hh:11]: For what it's worth, there was one "builder pattern" type situation where I did need to start with the empty set as an identity for merging proper ones in. So I dont really mind that there is an escape hatch. Of course yes the best thing to do would be to use a separate type for this one instance and make it ironclad.

File: src/libstore/outputs-spec.hh
Comment (Ericson2314) [reply src/libstore/outputs-spec.hh:28]: https://en.cppreference.com/w/cpp/utility/variant/monostate It gives me all the comparison functions and other such things I need.

File: src/libstore/outputs-spec.hh
Comment (Ericson2314) [reply src/libstore/outputs-spec.hh:32]: See https://github.com/NixOS/nix/issues/7479 we have an issue discussing this pattern. I am for it as the best that can be done for Sum types until we switch to Rust.

File: src/libstore/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/outputs-spec.cc]: It's gone!

File: src/libstore/outputs-spec.cc
Comment (Ericson2314) [reply src/libstore/outputs-spec.cc]: Done.

File: src/libstore/outputs-spec.hh
Comment (Ericson2314) [reply src/libstore/outputs-spec.hh]: Done.

File: src/libstore/outputs-spec.hh
Comment (roberth) [reply src/libstore/outputs-spec.hh:31]: It seems like that would be an appropriate use of `monostate`. It is a small class with a clear purpose, and by subclassing we make its meaning explicit.

What we're really doing here is declaring an algebraic data type, in a verbose, C++ way. Maybe that's why I don't see a problem, or may that's why there won't be a problem. Or maybe I'm misreading @tfc who isn't saying that this will be a problem per se and it isn't quite related to the comment about subclassing stl containers.

**Review by roberth — APPROVED** [review]
Looks really good, especially with the new tests, thanks @Ericson2314!
I'm also approving of using the established algebraic data type pattern, considering that it is better to discuss it as part of https://github.com/NixOS/nix/issues/7479. This new usage of it would not constitute any significant tech debt, as already a number of ADTs would have to be migrated to a different pattern.

**Comment by roberth** [comment]
Merged to unblock further work. The remaining discussion wasn't critical and can be continued in #7479

**Comment by edolstra** [comment]
@Ericson2314 Looks like this broke `build.aarch64-linux`, see https://hydra.nixos.org/build/205195243.

**Comment by Ericson2314** [comment]
@edolstra @roberth told me about the problem and I made #7604 which I hope fixes it? Though I wasn't able to test it.

**Comment by roberth** [comment]
I suspect that this PR has caused the regression:
 - #7655

**Pull request merged**
